quantity/core.py

```diff
@@ -0,0 +1,255 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import annotations
+
+import operator
+from dataclasses import dataclass, replace
+from typing import TYPE_CHECKING
+
+import array_api_compat
+import astropy.units as u
+import numpy as np
+from astropy.units.quantity_helper.converters import UFUNC_HELPERS
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+DIMENSIONLESS = u.dimensionless_unscaled
+
+PYTHON_NUMBER = float | int | complex
+
+
+def has_array_namespace(arg):
+    try:
+        array_api_compat.array_namespace(arg)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def get_value_and_unit(arg, default_unit=None):
+    # HACK: interoperability with astropy Quantity.  Have protocol?
```

Comment:
Agreed. Something like

```python
@runtime_checkable
class QuantityAPI(Protocol):
    value: Array
    unit: Unit
    
    def to(self, ...): ...
```

Then this function becomes

```python
def get_value_and_unit(arg, default_unit=None):
    return (arg.value, arg.unit) if isinstance(arg, QuantityAPI) else (arg, default_unit)
```

Reply:
I would like to actually raise an error on having a `.unit` attribute and not `.value` (e.g., how do we deal with `Column`?)

Reply:
p.s. There's part of me that still would like to avoid `Quantity` having anything to do with unit conversion - that can be up to the unit. E.g., `q.to_value(unit)` could become `q // unit` or `q >> unit` or so (somewhat partial to `//` myself, since it is close - `q1 // q2` already returns a dimensionless quantity, so all one really needs is to avoid acting on the "make integer" part).

Reply:
As in not have `.to()` and `.to_value`?

Reply:
Like in https://unxt.readthedocs.io/en/latest/#converting-units

![CleanShot 2024-11-26 at 15 11 53@2x](https://github.com/user-attachments/assets/8eb3bf28-4de5-416e-accc-8b771ddf9c7d)

(based on the Julia paradigm)

Reply:
OK to postpone this until we have a clearer idea what exactly the `Protocol` will be?

Reply:
Yes. Should we open an Issue?